File: Dart/SelectionSort.dart
void main() {
  List<Comparable> numbers = [5, 1, 4, 2, 8];
  List<Comparable> letters = ['q', 'a', 'd', 'e', 'h'];
  show("Old Array", numbers);
  sort(numbers);
  show("New Array", numbers);
}

// Selection Sort
void sort(List<Comparable> a) {
  int N = a.length;
  for (int i = 0; i < N; i++) {
    int min = i;
    for (int j = i + 1; j < N; j++) {
      if (isLess(a[j], a[min])) {
        min = j;
      }
    }
    // exchange a[i] and a[j]
    exchange(a, i, min);
  }
}

bool isLess(Comparable v, Comparable w) {
  return (v.compareTo(w) < 0);
}

void exchange(List a, int i, int j) {
  var swap = a[i];
  a[i] = a[j];
  a[j] = swap;
}

show(String text, List<Comparable> a) {
  print(text);
  for (int i = 0; i < a.length; i++) {
    print(a[i]);
  }
}

File: Rust/SelectionSort.rs
/// Implementation of Selection sort in rust
pub fn selection_sort<T: Ord>(arr: &mut [T]) {
    let len = arr.len();
    for left in 0..len {
        let mut smallest = left;
        for right in (left + 1)..len {
            if arr[right] < arr[smallest] {
                smallest = right;
            }
        }
        arr.swap(smallest, left);
    }
}

fn main() {
    println!("Sort numbers ascending");
    
    let mut numbers = [4, 65, 2, -31, 0, 99, 2, 83, 782, 1];
    println!("Before: {:?}", numbers);
    
    selection_sort(&mut numbers);
    println!("After:  {:?}\n", numbers);

    println!("Sort strings alphabetically");
    
    let mut strings = ["beach", "hotel", "airplane", "car", "house", "art"];
    println!("Before: {:?}", strings);
    
    selection_sort(&mut strings);
    println!("After:  {:?}\n", strings);
}
